pss-migration-orchestration/README.md

```diff
@@ -0,0 +1,86 @@
+- Generate Kyverno PolicyExceptions for customer administrators to manage. Instead of directly applying the Kyverno PolicyExceptions, we considered generating them as an alternative type with an identical spec, which cluster administrators could then store or distribute to workload teams as they choose. This would solve the immediate adoption limitations. However, to avoid hitting this issue again in the future, we hope to do slightly better.
+- Generate recommended generic exceptions for customer administrators to manage. Instead of directly creating the Kyverno PolicyExceptions, we instead generate an intermediate resource which customers can store or distribute to workload teams, and which separates the cluster admin's intent from the underlying implementation. This is the approach described in the rest of this RFC.
+
+## :exclamation: Risks
+
+1. The generated exceptions could be over-permissive. We actually expect this, and are working on a separate initiative to tighten down over-permissive exceptions in the future. We believe we can undertake this work only for purposes of migration, given that the generated exceptions will represent the current state of clusters secured with PSPs. By definition, the presence of a workload means that the resource has been approved and any current risk has been accepted by the customer, since their currently configured policy allows it. In the future, when PSS is the standard, we may not be able to make this assumption when changing policy implementations. The proposed process keeps customer admins in the loop, and they could elect to honor generated exceptions only from certain clusters, or from certain environments.
+2. This implementation could be too complicated, or we are not fast enough at implementing it. This is entirely possible. We think about this two ways. 1. There is upfront process work to be done by the customer, and without the automation recommending exceptions, it is unlikely that customers move very fast anyway. Any work done while we are implementing this capability is not wasted, and the recommendations can be used to accelerate adoption later on. 2. taking the time now to get customers directly to what we expect to be the future is preferable to trying to get them to repeat this effort again later.
+3. Probably others, we're moving pretty fast.
+
+## :feet: Implementation
+
+> [!NOTE]
+> This implementation includes new resource types outlined in a parallel RFC. We will only implement the minimum necessary to carry out the migration now.
```

Comment:
I actually agree that building a bunch of Giant Swarm abstractions is not ideal. I would be in favor of using any community-provided option to meet the same needs. However, there currently isn't such a thing.

> Kyverno is already complicated enough to understand...

Important to keep in mind: the idea is customers only touch these APIs when they need to deviate from platform defaults, or want to build automation on top. If a customer wants to create a _new_ policy that Giant Swarm does not manage, they would still do that by creating a normal Kyverno policy.

> a simplified migration plan could look like this
> Customer gets set up for automated reports

Unfortunately, it just isn't that simple.
Specifically:
Where do the reports go? Does each team get them for their own components? In many/most companies, the team won't get to give themselves exceptions like Giant Swarm does. So, do all the reports go to the security team? The platform team? Who actually creates all the exceptions? Do they know how to do that?

Multiple current customers have been working on this topic for a year already and struggling to answer those questions, which is basically why this is necessary at all.

Giant Swarm can move very quickly, but we can't move customers very quickly when it involves lots of change, especially to workloads (which often security policies do require).

Without some abstractions we will need customers to do additional high-effort migrations in the future.

Reply:
> Where do the reports go?

Fully agree with this organizational issue. If we add customizability to the reporting feature, it can also become a mess to maintain. Maybe a _very actionable_ dashboard could be a good start so customers get to know which workloads would fail. They might anyway build their own problem reporting in the long run, also very much depending on their tooling – for example Argo CD may show such deployment errors, and customers could just™️ alert based on Argo CD's app deployment health as a high-level status.